lib/api/ldp/ldp-request.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const HttpError = require('standard-http-error')
+
+class LdpRequest {
+  /**
+   * @param options {object}
+   *
+   * @param options.host {SolidHost} Host / config object
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.req {HttpRequest} Save the Express req object, useful
+   *   to access headers/preferences, writeStream for body, etc.
+   *
+   * @param [options.authn=null] {object} Authentication object, contains WebID
+   *   string, as well as any bearer credentials/tokens. Needed for
+   *   authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.parsedBody] {?} Request body, parsed when appropriate
+   *   (when content-type is parse-able)
+   */
+  constructor (options) {
+    this.host = options.host
+    this.target = options.target
+    this.req = options.req
+    this.authn = options.authn
+    this.parsedBody = options.parsedBody
+  }
+
+  static from ({req, host}) {
+    let request
+
+    switch (req.method) {
+      case 'head':
```

Comment:
Uppercase?

But also, I don't really like the `switch`. There should just be a `handlers` map somewhere, such that we can simply do `handler = handlers[req.method]`.

Reply:
Aren't Express request methods in lowercase?

I was torn between doing a `switch()` statement and a handlers map. 

I think I went with it for those cases (like with the GET method) where additional logic may need to be done before figuring out a subclass, so that a simple map would not suffice.

Any particular reason for disliking switch, out of curiosity?

Reply:
> Keeping specialized knowledge in one place. The class implementing the switch currently needs knowledge on the internals of the constructors it creates, spreading out this specific knowledge (especially if adding that additional logic).

So I think that's my point - the `switch()` statement *enables* specialized knowledge (as in the case of the GET operation - though that's not clear from the current code). (Though notice that it's not specialized knowledge of constructors, the subclass `from()` methods hide that knowledge).

Here, let me rename the `Ldp*Request` references to `Ldp*Operation`, to make the discussion clearer. (And take out the extra constructor params to LdpApi, as discussed). And add the illustrative code to the GET section of the `switch()`.

Reply:
> So I think that's my point - the switch() statement enables specialized knowledge (as in the case of the GET operation - though that's not clear from the current code).

Yes, but it spreads it out across two places: the `switch` statement and the `Ldp*Request` constructor. The class containing the `switch` has to know about each of the subclasses. That's knowledge it doesn't need, and ties the implementation of this class to each of the subclasses. That's coupling that can be avoided.

Reply:
> Yes, but it spreads it out across two places: the switch statement and the Ldp*Request constructor.

No prob. Lemme replace the switch() with a mapper, let's see how it looks.